nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java

```diff
@@ -0,0 +1,203 @@
+    private final AtomicReference<Boolean> stopping = new AtomicReference<>(false);
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        this.descriptors = ImmutableList.of(SOURCE_TYPE, AGENT_NAME, SOURCE_NAME, FLUME_CONFIG);
+        this.relationships = ImmutableSet.of(SUCCESS);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @OnScheduled
+    public void onScheduled(final SchedulingContext context) {
+        try {
+            stopping.set(false);
+            source = SOURCE_FACTORY.create(
+                context.getProperty(SOURCE_NAME)
+                .getValue(),
+                context.getProperty(SOURCE_TYPE)
+                .getValue());
+
+            String flumeConfig = context.getProperty(FLUME_CONFIG)
+                .getValue();
+            String agentName = context.getProperty(AGENT_NAME)
+                .getValue();
+            String sourceName = context.getProperty(SOURCE_NAME)
+                .getValue();
+            Configurables.configure(source,
+                getFlumeSourceContext(flumeConfig, agentName, sourceName));
+
+            if (source instanceof PollableSource) {
+                source.setChannelProcessor(new ChannelProcessor(
+                    new NifiChannelSelector(pollableSourceChannel)));
+                source.start();
+            }
+        } catch (Throwable th) {
+            getLogger()
+                .error("Error creating source", th);
+            throw Throwables.propagate(th);
+        }
+    }
+
+    @OnUnscheduled
+    public void unScheduled() {
+        stopping.set(true);
+        if (source instanceof PollableSource) {
+            source.stop();
+        } else {
+            EventDrivenSourceRunner runner = runnerRef.get();
+            if (runner != null) {
+                runner.stop();
+                runnerRef.compareAndSet(runner, null);
+            }
+
+            NifiSessionFactoryChannel eventDrivenSourceChannel = eventDrivenSourceChannelRef.get();
+            if (eventDrivenSourceChannel != null) {
+                eventDrivenSourceChannel.stop();
+                eventDrivenSourceChannelRef.compareAndSet(eventDrivenSourceChannel, null);
+            }
+        }
+    }
+
+    @OnStopped
+    public void stopped() {
+        sessionFactoryRef.set(null);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+        if (source instanceof PollableSource) {
+            super.onTrigger(context, sessionFactory);
+        } else if (source instanceof EventDrivenSource) {
+            ProcessSessionFactory old = sessionFactoryRef.getAndSet(sessionFactory);
+            if (old == null) {
```

Comment:
I based this on how the `ListenUdp` processor is written. I'm happy to throw an exception if it changes out from under us as that should only happen after the processor is stopped and started again and the `@OnStopped` method sets it back to `null`.